File: workers/javascript/sandbox.js
const vm = require('vm');

function runCode(code, inputData) {
    const logs = [];
    let status = 'SUCCESS';
    let output = '';

    // Create a custom console that captures logs
    const customConsole = {
        log: (...args) => logs.push(args.map(String).join(' ')),
        error: (...args) => logs.push('[ERROR] ' + args.map(String).join(' ')),
        warn: (...args) => logs.push('[WARN] ' + args.map(String).join(' ')),
        info: (...args) => logs.push(args.map(String).join(' ')),
    };

    // Sandbox context with limited globals
    const context = {
        event: inputData,
        console: customConsole,
        JSON: JSON,
        Math: Math,
        Date: Date,
        Array: Array,
        Object: Object,
        String: String,
        Number: Number,
        Boolean: Boolean,
        parseInt: parseInt,
        parseFloat: parseFloat,
        isNaN: isNaN,
        isFinite: isFinite,
    };

    try {
        vm.createContext(context);

        // Execute user code to define handler function
        vm.runInContext(code, context, {
            timeout: 30000, // 30 second timeout
            displayErrors: true,
        });

        // Check if handler function exists
        if (typeof context.handler !== 'function') {
            status = 'ERROR';
            output = "No 'handler(event)' function defined in code.";
        } else {
            // Call handler function
            const result = context.handler(inputData);
            output = result !== undefined ? JSON.stringify(result) : 'undefined';
        }
    } catch (err) {
        status = 'ERROR';
        output = err.message;
        logs.push(err.stack || err.toString());
    }

    return {
        status,
        output,
        logs: logs.join('\n'),
    };
}

module.exports = { runCode };

File: workers/ruby/sandbox.rb
require 'json'
require 'timeout'
require 'stringio'

module Sandbox
  TIMEOUT_SECONDS = 30

  def self.run_code(code, input_data)
    logs = []
    status = "SUCCESS"
    output = ""

    # Create a new binding for isolated execution
    sandbox_binding = create_sandbox_binding(input_data, logs)

    begin
      Timeout.timeout(TIMEOUT_SECONDS) do
        # Execute user code in sandbox
        eval(code, sandbox_binding, "user_code.rb")

        # Check if handler method is defined and call it
        if sandbox_binding.local_variable_defined?(:handler_result)
          result = sandbox_binding.local_variable_get(:handler_result)
          output = result.to_json
        else
          # Try to call handler function if defined
          begin
            handler_proc = sandbox_binding.eval("method(:handler)")
            result = handler_proc.call(input_data)
            output = result.to_json
          rescue NameError
            status = "ERROR"
            output = "No 'handler' method defined in code. Please define: def handler(event) ... end"
          end
        end
      end
    rescue Timeout::Error
      status = "TIMEOUT"
      output = "Execution timed out after #{TIMEOUT_SECONDS} seconds"
    rescue SyntaxError => e
      status = "ERROR"
      output = "Syntax error: #{e.message}"
      logs << e.backtrace&.first(5)&.join("\n")
    rescue => e
      status = "ERROR"
      output = "Runtime error: #{e.message}"
      logs << e.backtrace&.first(10)&.join("\n")
    end

    [status, output, logs.flatten.compact.join("\n")]
  end

  private

  def self.create_sandbox_binding(input_data, logs)
    # Create isolated binding with limited access
    sandbox = Object.new

    # Define puts/print to capture logs
    sandbox.define_singleton_method(:puts) do |*args|
      logs << args.map(&:to_s).join(' ')
    end

    sandbox.define_singleton_method(:print) do |*args|
      logs << args.map(&:to_s).join('')
    end

    sandbox.define_singleton_method(:p) do |*args|
      logs << args.map(&:inspect).join(', ')
    end

    # Provide event data
    sandbox.define_singleton_method(:event) { input_data }

    sandbox.instance_eval { binding }
  end
end
